internal/kafka/internal/services/kafka.go

```diff
@@ -179,6 +181,56 @@ func (k *kafkaService) capacityAvailableForRegionAndInstanceType(instTypeRegCapa
 	return instTypeRegCapacity == nil || count <= int64(*instTypeRegCapacity), nil
 }
 
+func (k *kafkaService) GetAvailableSizesInRegion(criteria *FindClusterCriteria) ([]string, *errors.ServiceError) {
+	supportedInstanceTypes := k.kafkaConfig.SupportedInstanceTypes.Configuration
+	it, err := supportedInstanceTypes.GetKafkaInstanceTypeByID(criteria.SupportedInstanceType)
+	if it == nil {
+		errMsg := fmt.Sprintf("unable to get available sizes in region: failed to find instance type: %s", err.Error())
+		glog.Infoln(errMsg)
+		return []string{}, errors.New(errors.ErrorGeneral, errMsg)
+	}
+
+	// The kafka size list configuration must always be ordered starting with the smallest unit.
+	// The following finds the largest Kafka size that is still available in this region. Anything smaller than this
+	// size will also be considered as available to create with the remaining capacity.
+	for i := len(it.Sizes) - 1; i >= 0; i-- {
+		kafka := &dbapi.KafkaRequest{
+			CloudProvider: criteria.Provider,
+			Region:        criteria.Region,
+			InstanceType:  criteria.SupportedInstanceType,
+			MultiAZ:       criteria.MultiAZ,
+			SizeId:        it.Sizes[i].Id,
+		}
+
+		// Check against region limits
+		hasCapacity, err := k.HasAvailableCapacityInRegion(kafka)
+		if err != nil {
+			errMsg := fmt.Sprintf("unable to get available sizes in region: failed to check region capacity: %s", err.Error())
+			glog.Infoln(errMsg)
+			return []string{}, errors.New(errors.ErrorGeneral, errMsg)
```

Comment:
no need to create an empty array: `nil` can be safely returned here

Reply:
I tried this at the start but in the response I get the following: 
```
{
    "instance_type": "standard",
    "max_capacity_reached": true,
    "available_sizes": null
},
```

I could be missing something from the openapi side of things though. In the openapi docs, since this property is required, I can't set a default value. `nullable` is also false, but maybe cause it has the type array it can be? :thinking: 

I could set the value of the `available_sizes` to an empty slice outside of this function though if it returns nil. wdyt?

Reply:
+1 on creating an empty array to avoid the `null` issue. 
I think we are doing that already in some parts of the codebase to avoid not sending the field at all or setting it to `null`.

Reply:
But here we are returning an error. If this null arrives to the response, means the error is ignored somewhere...

Reply:
I actually have a question around the error handling here. So an error will be returned in `GetAvailableSizesInRegion()` with the following conditions:
- provider/region/instance type/size is not supported
   - if any are unsupported, we should just return an empty array.
- error occurred when trying to retrieve Kafkas from the db.
   - should we return a 500 in this case?

It's hard to differentiate between the two atm. We will probably need to refactor [FindCluster](https://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/blob/main/internal/kafka/internal/services/cluster_placement_strategy.go) to return a serviceError so we can differentiate between the two. 

what do you guys think? If it's the case that we don't really need to return an err to the client then i'm ok with removing the returned error here and just log it.

Reply:
This is an old and very discussed topic.

I'm not too expert in `go` best practice, so I will try to make a generic reasoning: when I invoke a `Find` function, the function is expected to search for an object. The result of search could be a `object found` or `object not found`.

Not finding the object is a normal behaviour for a `find` function, so no error should be returned.
In Java or C++ I would use an `optional` object to avoid using a null return.

The best I can thing of with `go` could be to change the `FindCluster` so that it returns 3 values:

`func FindCluster(kafka *dbapi.KafkaRequest) (*api.Cluster, bool, error)`

so when we call it we do:

```go
cluster, found, error := FindCluster(req)
```

Reply:
It makes sense to refactor it that way. There's still a question of do we return an error back to the user/client for db errors though? If that's the case, that refactor needs to be done along with this PR. 

If not, that could be an enhancement in a follow on PR. Since we don't need to return the error back to the user, we can remove the returned err in `GetAvailableSizesInRegion`.

Reply:
Thinking about this again, I think we should probably return the 500 for the db errors. It could be misleading to report there's no capacity left even though that may not be true and there is actually a problem on the server side retrieving this list.

Reply:
Totally agree

Reply:
@ziccardi updated this now. 

FindCluster should now return a ServiceError. Had to do it this way than what was described above because we already kind of have the `found` when the cluster returned is nil. What we really need is a way to differentiate between the different type of errors returned by FindCluster since in the GET cloud provider regions handler, we need to ignore any non General errors. 

The GetAvailableSizes() should now return nil instead of creating an empty array on return.

Reply:
> we already kind of have the found when the cluster returned is nil

It can certainly be done that way. The only drawback is that if for any reason the caller won't check for nil, he will get a panic. On the other side, if you get a boolean you intentionally have to ignore it.